### ISSUE TITLE
Discussion: fallback import searches that add .dhall

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-lang#680]
Currently the Prelude has a lot of files that have no extension. This means for a lot of editors, they will not by default show any syntax highlighting without extra work. However, the lack of extension makes it very "clean" to reference Dhall files - you don't have the baggage of the file extension. As a compromise so both sides get the best experience, what do we think about having a fallback import that tries the same import, but with `.dhall`? That is:

```dhall
./Hello
```

Will first try and import `./Hello` as before, but if `./Hello` doesn't exist, it will try again with `./Hello.dhall`.

This is a very half baked idea, just throwing it out to see what others think.

**Comment by Gabriel439** [comment]
Yeah, this is actually something I've been meaning to fix for a while but keep forgetting to do so.  The solution I had in mind was that the actual code for all Prelude utilities would live in `.dhall` files and the prefix-free versions would just import their `.dhall` counterparts (for backwards compatibility)

**Comment by ocharles** [comment]
Ok, that fixes it at the individual level, with the cost of requiring a lot more busy work. It's certainly an option.

**Comment by singpolyma** [comment]
This is popular in many other scripting languages, so I'd consider it an option.

It does have overhead for HTTP(S) imports though.

**Comment by Gabriel439** [comment]
Oh, I misunderstood the proposal.  So if the goal is to automatically fall back with an extension for *all* imports then I'm mildly opposed, for the same reasons that I removed support for falling back to a default import for directories:

* It's more magic that users have to understand to use the language
* It deteriorates error messages

**Comment by ocharles** [comment]
Yes, these are exactly my concerns, but the ergonomic improvement is not insignificant.

**Comment by ocharles** [comment]
I suppose if you just do imports "up front" as you would in other languages it's not so bad:

```dhall
let Hello = ./Hello.dhall
```

and I think with pattern matching in `let`s, you can start to model a bit of an import system:

```dhall
let { foo, bar } = ./package.dhall
```

**Comment by singpolyma** [comment]
> let { foo, bar } = ./package.dhall

I keep meaning to write a proposal for full-unpacking

    let {..} = ...

syntax. Even named matching as you describe would be pretty nice to have for imports.

**Comment by ocharles** [comment]
See https://github.com/dhall-lang/dhall-lang/issues/74

**Comment by Gabriel439** [comment]
Yeah, I almost never do inline imports.  I usually put them at the top of the file using `let` expressions as if they were traditional imports

**Comment by Profpatsch** [comment]
I’m against this change (auto-append `.dhall`), for the simple reason that I can’t do `gf` in vim anymore to jump to the imported file.
